test/unit/test-multipolicy.cpp

```diff
@@ -44,51 +44,87 @@
 /// Source file containing tests for basic multipolicy operation
 ///
 
+#include <cstddef>
 #include "gtest/gtest.h"
 
+struct mp_tag1 {
+};
+struct mp_tag2 {
+};
+struct mp_tag3 {
+};
+
+struct mp_test_body;
+namespace RAJA
+{
+namespace impl
+{
+template <typename Iterable>
+void forall(const mp_tag1 &p, Iterable &&iter, mp_test_body const &body)
+{
+  body(p, iter.size());
+}
+template <typename Iterable>
+void forall(const mp_tag2 &p, Iterable &&iter, mp_test_body const &body)
+{
+  body(p, iter.size());
+}
+template <typename Iterable>
+void forall(const mp_tag3 &p, Iterable &&iter, mp_test_body const &body)
+{
+  body(p, iter.size());
+}
+}
+}
+
+// NOTE: this *must* be after the above to work
 #include "RAJA/RAJA.hpp"
 
+struct mp_test_body {
+  void operator()(mp_tag1 const &, std::size_t size) const
+  {
+    ASSERT_LT(size, std::size_t{100});
+  }
+  void operator()(mp_tag2 const &, std::size_t size) const
+  {
+    ASSERT_GT(size, std::size_t{99});
+  }
+  void operator()(mp_tag3 const &, std::size_t size) const
+  {
+    ASSERT_GT(size, std::size_t{10});
+    ASSERT_LT(size, std::size_t{99});
+  }
+};
+
+
 TEST(MultiPolicy, basic)
 {
-  auto mp =
-      RAJA::make_multi_policy<RAJA::seq_exec, RAJA::omp_parallel_for_exec>(
-          [](const RAJA::RangeSegment& r) {
-            if (r.size() < 100) {
-              return 0;
-            } else {
-              return 1;
-            }
-          });
-  RAJA::forall(mp, RAJA::RangeSegment(0, 5), [](RAJA::Index_type) {
-    ASSERT_EQ(omp_get_num_threads(), 1);
-  });
-  RAJA::forall(mp, RAJA::RangeSegment(0, 101), [](RAJA::Index_type) {
-    ASSERT_TRUE(omp_get_num_threads() > 1);
-  });
+  auto mp = RAJA::make_multi_policy<mp_tag1, mp_tag2>(
```

Comment:
I like the structure of this, but can you comment this file some to explain what's expected and how the assertions relate to the tests. (I can see what's going on, but it would be useful to have it recorded for posterity).